Add v3 and v4 encodings, using mochilo

spraints opened the pull request:
This isn't quite done yet, but I wanted to open up this PR to capture the progress so far. This adds a new BERT encoding (v3) that's based on msgpack. This is much faster, especially for long arrays.

```
                                     user     system      total        real
BERT v1 tiny                     0.020000   0.000000   0.020000 (  0.019957)
BERT v1 small                    0.150000   0.000000   0.150000 (  0.141066)
BERT v1 large                    0.710000   0.000000   0.710000 (  0.723820)
BERT v1 complex                  2.950000   0.000000   2.950000 (  2.972712)
BERT v1 long array               3.870000   0.000000   3.870000 (  3.898899)
BERT v2 tiny                     0.010000   0.000000   0.010000 (  0.017890)
BERT v2 small                    0.160000   0.000000   0.160000 (  0.159581)
BERT v2 large                    0.720000   0.000000   0.720000 (  0.733793)
BERT v2 complex                  2.960000   0.010000   2.970000 (  2.975499)
BERT v2 long array               4.080000   0.000000   4.080000 (  4.106795)
BERT v3 tiny                     0.020000   0.000000   0.020000 (  0.024472)
BERT v3 small                    0.020000   0.000000   0.020000 (  0.025297)
BERT v3 large                    0.320000   0.020000   0.340000 (  0.337157)
BERT v3 complex                  0.090000   0.000000   0.090000 (  0.085884)
BERT v3 long array               0.100000   0.000000   0.100000 (  0.104439)
```

String encodings are currently not working. I've heard that bananapack (https://github.com/brianmario/mochilo) could solve this. I skimmed through the mochilo code, and I'm not sure that it'll work, because it lacks the [custom type stuff](https://github.com/spraints/bert/blob/5c88d933e09df9314ca1a0b6bffdd70edf6d629c/lib/bert/msgpack.rb).

cc @carlosmn @piki @vmg

spraints committed:
Benchmark dat msgpack

spraints committed:
bert_test: restructure to prep for v3 tests

spraints committed:
Make a basic v3(msgpack) encoder

spraints committed:
decode msgpack-based bert

spraints committed:
set up serialization of symbols

spraints committed:
do less processing on a msgpackable object

* skip 'convert', because the decomposition isn't interesting.
* register more types, so that msgpack can deal with all the types that
  BERT can

spraints committed:
benchmark the new BERT encoder

It's slightly slower than plain msgpack, but still much better than
older BERT.

                                     user     system      total        real
BERT v1 tiny                     0.020000   0.000000   0.020000 (  0.019957)
BERT v1 small                    0.150000   0.000000   0.150000 (  0.141066)
BERT v1 large                    0.710000   0.000000   0.710000 (  0.723820)
BERT v1 complex                  2.950000   0.000000   2.950000 (  2.972712)
BERT v1 long array               3.870000   0.000000   3.870000 (  3.898899)
BERT v2 tiny                     0.010000   0.000000   0.010000 (  0.017890)
BERT v2 small                    0.160000   0.000000   0.160000 (  0.159581)
BERT v2 large                    0.720000   0.000000   0.720000 (  0.733793)
BERT v2 complex                  2.960000   0.010000   2.970000 (  2.975499)
BERT v2 long array               4.080000   0.000000   4.080000 (  4.106795)
BERT v3 tiny                     0.020000   0.000000   0.020000 (  0.024472)
BERT v3 small                    0.020000   0.000000   0.020000 (  0.025297)
BERT v3 large                    0.320000   0.020000   0.340000 (  0.337157)
BERT v3 complex                  0.090000   0.000000   0.090000 (  0.085884)
BERT v3 long array               0.100000   0.000000   0.100000 (  0.104439)
JSON tiny                        0.010000   0.000000   0.010000 (  0.007604)
JSON small                       0.010000   0.000000   0.010000 (  0.014245)
JSON large                       2.450000   0.010000   2.460000 (  2.473644)
JSON complex                     0.170000   0.000000   0.170000 (  0.174540)
JSON long array                  0.350000   0.000000   0.350000 (  0.357081)
YAJL tiny                        0.010000   0.000000   0.010000 (  0.005566)
YAJL small                       0.020000   0.000000   0.020000 (  0.019177)
YAJL large                       1.060000   0.030000   1.090000 (  1.094342)
YAJL complex                     0.300000   0.000000   0.300000 (  0.307467)
YAJL long array                  0.140000   0.000000   0.140000 (  0.136261)
Ruby tiny                        0.000000   0.000000   0.000000 (  0.002919)
Ruby small                       0.010000   0.000000   0.010000 (  0.008211)
Ruby large                       0.000000   0.010000   0.010000 (  0.009826)
Ruby complex                     0.020000   0.000000   0.020000 (  0.013560)
Ruby long array                  0.080000   0.000000   0.080000 (  0.082170)
Msgpack tiny                     0.000000   0.000000   0.000000 (  0.002561)
Msgpack small                    0.000000   0.000000   0.000000 (  0.003771)
Msgpack large                    0.190000   0.000000   0.190000 (  0.195700)
Msgpack complex                  0.040000   0.010000   0.050000 (  0.044514)
Msgpack long array               0.070000   0.000000   0.070000 (  0.071905)

piki commented on bert.gemspec:
```diff
@@ -21,6 +21,9 @@ Gem::Specification.new do |s|
 
+    s.add_dependency "msgpack"
+
+    s.add_development_dependency "byebug"
```
Remove `byebug` before merging, I assume?

carlosmn commented:
> I skimmed through the mochilo code, and I'm not sure that it'll work, because it lacks the custom type stuff.

In that case, it sounds like we'll have to build it ourselves.

Since we already have `Encoder.convert` to convert the ruby objects into whatever we need to pass though, we should be able to return some `EncodedString` type and register our own type, which packs as a tuple of contents plus encoding. That should be enough to work around msgpack always trying to use its own type for strings, no?

carlosmn commented:
That seems to work out well enough. It's presumably worth not returning `EncodedString` if we see it's binary or UTF-8 or whatever msgpack can do itself.

```diff
diff --git a/lib/bert.rb b/lib/bert.rb
index 352016a..be499f7 100644
--- a/lib/bert.rb
+++ b/lib/bert.rb
@@ -30,3 +30,17 @@ require 'bert/decoder'
 def t
   BERT::Tuple
 end
+
+class EncodedString
+  def initialize(string)
+    @string = string
+  end
+
+  def raw
+    @string.byteslice(0..-1)
+  end
+
+  def encoding
+    @string.encoding
+  end
+end
diff --git a/lib/bert/encoder.rb b/lib/bert/encoder.rb
index 3e9bd0b..6e4bb77 100644
--- a/lib/bert/encoder.rb
+++ b/lib/bert/encoder.rb
@@ -19,7 +19,15 @@ module BERT
     #
     # Returns the converted Ruby object
     def self.convert(item)
-      return item if Encode.version == :v3
+      if Encode.version == :v3
+        case item
+        when String
+          return EncodedString.new(item)
+        else
+          return item
+        end
+      end
+
       case item
         when Hash
           pairs = []
diff --git a/lib/bert/msgpack.rb b/lib/bert/msgpack.rb
index 49ff96e..600844d 100644
--- a/lib/bert/msgpack.rb
+++ b/lib/bert/msgpack.rb
@@ -11,6 +11,9 @@ module BERT
     factory.register_type 0x03, Time,
       :packer => lambda { |time| [time.to_f].pack("G") },
       :unpacker => lambda { |raw| Time.at(raw.unpack("G").first) }
+    factory.register_type 0x04, EncodedString,
+      :packer => lambda { |s| [s.encoding.name, s.raw].pack("Z*a*") },
+      :unpacker => lambda { |raw| enc, str = raw.unpack("Z*a*"); str.force_encoding(enc); str }
     return factory
   end
 end
```

spraints committed:
Use a modified msgpack, so that String can be customized

This is slower by a lot.

< BERT v3 long array               0.080000   0.000000   0.080000 (  0.083473)
> BERT v3 long array               1.960000   0.000000   1.960000 (  1.981249)

spraints committed:
Make v4, using mochilo

mochilo is like msgpack, except that it deals with encodings in a
sane(r?) way. mochilo doesn't have a custom type registry, so I've added
one in the "custom-type-registry" branch, and I'm using that here.

This ended up being faster than the msgpack implementation, because
encoding doesn't have to pop up to ruby to deal with strings.

                                     user     system      total        real
BERT v2 tiny                     0.010000   0.000000   0.010000 (  0.015289)
BERT v2 small                    0.130000   0.000000   0.130000 (  0.126682)
BERT v2 large                    0.580000   0.000000   0.580000 (  0.592555)
BERT v2 complex                  2.880000   0.010000   2.890000 (  2.903082)
BERT v2 long array               4.200000   0.000000   4.200000 (  4.262165)

BERT v3 tiny                     0.020000   0.000000   0.020000 (  0.020252)
BERT v3 small                    0.020000   0.000000   0.020000 (  0.024444)
BERT v3 large                    0.700000   0.010000   0.710000 (  0.703152)
BERT v3 complex                  0.090000   0.000000   0.090000 (  0.100372)
BERT v3 long array               1.810000   0.010000   1.820000 (  1.827934)

BERT v4 tiny                     0.010000   0.000000   0.010000 (  0.013301)
BERT v4 small                    0.010000   0.000000   0.010000 (  0.012317)
BERT v4 large                    0.230000   0.060000   0.290000 (  0.292852)
BERT v4 complex                  0.070000   0.000000   0.070000 (  0.064578)
BERT v4 long array               0.110000   0.000000   0.110000 (  0.110842)

Msgpack tiny                     0.000000   0.000000   0.000000 (  0.001644)
Msgpack small                    0.010000   0.000000   0.010000 (  0.002568)
Msgpack large                    0.180000   0.000000   0.180000 (  0.192238)
Msgpack complex                  0.040000   0.000000   0.040000 (  0.041482)
Msgpack long array               0.070000   0.000000   0.070000 (  0.065986)

spraints committed:
remove byebug dependency

spraints commented:
Does the `EncodedString` thing work for strings that are nested in an array or hash?

I tried two other implementations: one using a modified version of msgpack that allows customizations of the String serializer (https://github.com/msgpack/msgpack-ruby/compare/msgpack:808fdeb...spraints:8a2f5bf), and one using a modified version of mochilo that allows custom types to be registered (https://github.com/brianmario/mochilo/compare/brianmario:ced75e5...spraints:458fcb5). I like the mochilo version better because it's faster, and I think it'll be easier to get a PR merged.

For now, the msgpack version is called "v3" and the mochilo version is called "v4", so that I can compare them side-by-side. Whichever one wins will end up being called "v3".

carlosmn commented:
>Does the EncodedString thing work for strings that are nested in an array or hash?

Good point. The code as posted won't do that (I guess we should add a test ;), we'd have to do something similar to the handling in actual BERT where we `item.map { |i| convert(i) }`.

If we can get mochilo to accept custom types, I do think that sounds better, as it saves us from handling the encoding in this gem and would let us re-use that wherever we want mgpack in the future.

spraints committed:
v4 (mochilo) -> v3

spraints committed:
Serialize symbol, time, regexp in the benchmark, too

                                     user     system      total        real
BERT v1 tiny                     0.010000   0.000000   0.010000 (  0.013280)
BERT v1 small                    0.110000   0.000000   0.110000 (  0.116110)
BERT v1 large                    0.630000   0.010000   0.640000 (  0.646742)
BERT v1 complex                  2.710000   0.000000   2.710000 (  2.738346)
BERT v1 long array               5.360000   0.000000   5.360000 (  5.391206)
BERT v2 tiny                     0.020000   0.000000   0.020000 (  0.018202)
BERT v2 small                    0.140000   0.000000   0.140000 (  0.133662)
BERT v2 large                    0.640000   0.010000   0.650000 (  0.660656)
BERT v2 complex                  2.520000   0.010000   2.530000 (  2.534638)
BERT v2 long array               5.460000   0.010000   5.470000 (  5.520119)
BERT v3 tiny                     0.010000   0.000000   0.010000 (  0.010693)
BERT v3 small                    0.010000   0.000000   0.010000 (  0.011390)
BERT v3 large                    0.220000   0.030000   0.250000 (  0.246068)
BERT v3 complex                  0.060000   0.000000   0.060000 (  0.055431)
BERT v3 long array               0.520000   0.000000   0.520000 (  0.522735)
JSON tiny                        0.000000   0.000000   0.000000 (  0.005991)
JSON small                       0.020000   0.000000   0.020000 (  0.010092)
JSON large                       2.200000   0.020000   2.220000 (  2.252311)
JSON complex                     0.160000   0.000000   0.160000 (  0.159742)
JSON long array                  0.450000   0.010000   0.460000 (  0.454021)
YAJL tiny                        0.010000   0.000000   0.010000 (  0.006357)
YAJL small                       0.020000   0.000000   0.020000 (  0.023327)
YAJL large                       0.960000   0.020000   0.980000 (  1.002778)
YAJL complex                     0.280000   0.030000   0.310000 (  0.317845)
YAJL long array                  0.750000   0.000000   0.750000 (  0.747592)
Ruby tiny                        0.000000   0.000000   0.000000 (  0.003547)
Ruby small                       0.010000   0.000000   0.010000 (  0.007403)
Ruby large                       0.010000   0.000000   0.010000 (  0.012739)
Ruby complex                     0.020000   0.000000   0.020000 (  0.015415)
Ruby long array                  0.030000   0.000000   0.030000 (  0.036519)

spraints committed:
Use brianmario/mochilo#19

The long array time was:
BERT v3 long array               0.520000   0.000000   0.520000 (  0.522735)

and now it is:
BERT v3 long array               0.190000   0.000000   0.190000 (  0.183106)

The full benchmark output:
                                     user     system      total        real
BERT v1 tiny                     0.010000   0.000000   0.010000 (  0.013550)
BERT v1 small                    0.120000   0.020000   0.140000 (  0.124408)
BERT v1 large                    0.520000   0.000000   0.520000 (  0.534007)
BERT v1 complex                  2.730000   0.010000   2.740000 (  2.758483)
BERT v1 long array               5.150000   0.030000   5.180000 (  5.202551)
BERT v2 tiny                     0.020000   0.000000   0.020000 (  0.019604)
BERT v2 small                    0.120000   0.000000   0.120000 (  0.121305)
BERT v2 large                    0.630000   0.000000   0.630000 (  0.630657)
BERT v2 complex                  2.590000   0.000000   2.590000 (  2.586704)
BERT v2 long array               5.390000   0.020000   5.410000 (  5.453956)
BERT v3 tiny                     0.000000   0.000000   0.000000 (  0.005028)
BERT v3 small                    0.000000   0.010000   0.010000 (  0.006714)
BERT v3 large                    0.220000   0.040000   0.260000 (  0.259619)
BERT v3 complex                  0.040000   0.000000   0.040000 (  0.040061)
BERT v3 long array               0.190000   0.000000   0.190000 (  0.183106)
JSON tiny                        0.000000   0.000000   0.000000 (  0.008714)
JSON small                       0.020000   0.000000   0.020000 (  0.013369)
JSON large                       2.280000   0.010000   2.290000 (  2.332599)
JSON complex                     0.150000   0.000000   0.150000 (  0.145547)
JSON long array                  0.530000   0.000000   0.530000 (  0.526635)
YAJL tiny                        0.000000   0.000000   0.000000 (  0.005164)
YAJL small                       0.020000   0.000000   0.020000 (  0.016937)
YAJL large                       0.940000   0.030000   0.970000 (  0.987986)
YAJL complex                     0.230000   0.010000   0.240000 (  0.233397)
YAJL long array                  0.710000   0.000000   0.710000 (  0.710261)
Ruby tiny                        0.000000   0.000000   0.000000 (  0.003004)
Ruby small                       0.000000   0.000000   0.000000 (  0.005335)
Ruby large                       0.020000   0.000000   0.020000 (  0.010745)
Ruby complex                     0.010000   0.000000   0.010000 (  0.014781)
Ruby long array                  0.030000   0.010000   0.040000 (  0.033941)

spraints committed:
C implementation for v3

Even though the C implementation calls back to Ruby when it runs
mochilo, the v3 bert decoder is still faster in C than Ruby.

BERT_TEST_IMPL=C bundle exec ruby bench/bench.rb:

                                     user     system      total        real
BERT v1 tiny                     0.000000   0.000000   0.000000 (  0.009276)
BERT v1 small                    0.070000   0.000000   0.070000 (  0.091471)
BERT v1 large                    0.310000   0.030000   0.340000 (  0.342087)
BERT v1 complex                  1.130000   0.010000   1.140000 (  1.144703)
BERT v1 long array               2.240000   0.010000   2.250000 (  2.257003)
BERT v2 tiny                     0.010000   0.000000   0.010000 (  0.008312)
BERT v2 small                    0.060000   0.000000   0.060000 (  0.060999)
BERT v2 large                    0.320000   0.020000   0.340000 (  0.346957)
BERT v2 complex                  1.140000   0.000000   1.140000 (  1.135720)
BERT v2 long array               2.020000   0.000000   2.020000 (  2.037739)
BERT v3 tiny                     0.000000   0.000000   0.000000 (  0.003123)
BERT v3 small                    0.000000   0.000000   0.000000 (  0.003621)
BERT v3 large                    0.240000   0.040000   0.280000 (  0.289285)
BERT v3 complex                  0.040000   0.000000   0.040000 (  0.036944)
BERT v3 long array               0.170000   0.000000   0.170000 (  0.170969)

BERT_TEST_IMPL=Ruby bundle exec ruby bench/bench.rb:

                                     user     system      total        real
BERT v1 tiny                     0.010000   0.000000   0.010000 (  0.013944)
BERT v1 small                    0.130000   0.000000   0.130000 (  0.124755)
BERT v1 large                    0.560000   0.010000   0.570000 (  0.571472)
BERT v1 complex                  2.750000   0.000000   2.750000 (  2.758103)
BERT v1 long array               5.440000   0.000000   5.440000 (  5.476571)
BERT v2 tiny                     0.020000   0.000000   0.020000 (  0.017456)
BERT v2 small                    0.140000   0.000000   0.140000 (  0.136073)
BERT v2 large                    0.640000   0.000000   0.640000 (  0.648367)
BERT v2 complex                  2.610000   0.000000   2.610000 (  2.627479)
BERT v2 long array               5.490000   0.010000   5.500000 (  5.519414)
BERT v3 tiny                     0.010000   0.000000   0.010000 (  0.003783)
BERT v3 small                    0.000000   0.000000   0.000000 (  0.005775)
BERT v3 large                    0.240000   0.010000   0.250000 (  0.251264)
BERT v3 complex                  0.040000   0.000000   0.040000 (  0.037354)
BERT v3 long array               0.180000   0.000000   0.180000 (  0.183192)

spraints commented:
This is now using brianmario/mochilo#19, which handles the extra types that BERT needs in C, and v3 is now even faster. (Note that I changed the benchmark a little bit, because the 45s test cases were annoying to wait for.) Here are the times for the BERT's Ruby decoders:

```
                                     user     system      total        real
BERT v1 tiny                     0.010000   0.000000   0.010000 (  0.013944)
BERT v1 small                    0.130000   0.000000   0.130000 (  0.124755)
BERT v1 large                    0.560000   0.010000   0.570000 (  0.571472)
BERT v1 complex                  2.750000   0.000000   2.750000 (  2.758103)
BERT v1 long array               5.440000   0.000000   5.440000 (  5.476571)
BERT v2 tiny                     0.020000   0.000000   0.020000 (  0.017456)
BERT v2 small                    0.140000   0.000000   0.140000 (  0.136073)
BERT v2 large                    0.640000   0.000000   0.640000 (  0.648367)
BERT v2 complex                  2.610000   0.000000   2.610000 (  2.627479)
BERT v2 long array               5.490000   0.010000   5.500000 (  5.519414)
BERT v3 tiny                     0.010000   0.000000   0.010000 (  0.003783)
BERT v3 small                    0.000000   0.000000   0.000000 (  0.005775)
BERT v3 large                    0.240000   0.010000   0.250000 (  0.251264)
BERT v3 complex                  0.040000   0.000000   0.040000 (  0.037354)
BERT v3 long array               0.180000   0.000000   0.180000 (  0.183192)
```

Here are the times for BERT's C decoders:

```
                                     user     system      total        real
BERT v1 tiny                     0.000000   0.000000   0.000000 (  0.009276)
BERT v1 small                    0.070000   0.000000   0.070000 (  0.091471)
BERT v1 large                    0.310000   0.030000   0.340000 (  0.342087)
BERT v1 complex                  1.130000   0.010000   1.140000 (  1.144703)
BERT v1 long array               2.240000   0.010000   2.250000 (  2.257003)
BERT v2 tiny                     0.010000   0.000000   0.010000 (  0.008312)
BERT v2 small                    0.060000   0.000000   0.060000 (  0.060999)
BERT v2 large                    0.320000   0.020000   0.340000 (  0.346957)
BERT v2 complex                  1.140000   0.000000   1.140000 (  1.135720)
BERT v2 long array               2.020000   0.000000   2.020000 (  2.037739)
BERT v3 tiny                     0.000000   0.000000   0.000000 (  0.003123)
BERT v3 small                    0.000000   0.000000   0.000000 (  0.003621)
BERT v3 large                    0.240000   0.040000   0.280000 (  0.289285)
BERT v3 complex                  0.040000   0.000000   0.040000 (  0.036944)
BERT v3 long array               0.170000   0.000000   0.170000 (  0.170969)
```

spraints committed:
Bump mochilo

spraints committed:
Use a varying mochilo version in CI

spraints committed:
mochilo v1 == bert v3, mochilo v2 == bert v4

spraints committed:
Update dependencies so that a new enough mochilo is used

In the gemspec, limit the mochilo version to something no older than
1.3 (brianmario/mochilo#20) for the 1.x series, and no older than 2.1
(brianmario/mochilo#19) for the 2.x series.

In the main Gemfile and in the v2 CI gemfile, bundle from github.com
because there hasn't been a gem cut yet.

In the v1 CI gemfile, bundle from the PR branch.

spraints committed:
brianmario/mochilo#20 is merged

spraints commented:
This is ready for review.

carlosmn reviewed (approved):
Makes sense to me. The only think I'm wondering about is how we'd end up shipping the mochilo gem. We'd need to vendor it in the app, no?

spraints commented:
In the app yes, I'll bundle it. Here I was planning to leave it bundled via git, but I could put the gem files in vendor/cache. The *right* answer is probably to publish the gem updates.

piki commented:
Looks fine to me.  I'm curious, why are there both C and pure-Ruby implementations of the decoder?

spraints commented:
> I'm curious, why are there both C and pure-Ruby implementations of the decoder?

I don't know. That goes back to a commit 8y ago that says "remove reliance on [erlectricity](https://github.com/mojombo/erlectricity)" (41503465479e8762916d6997d91639f0d7308a13). 

> Only those with write access to this repository can merge pull requests.

cc @charliesome @brianmario @carlosmn This PR is ready to be merged, but I don't have a green merge button.
